src/views/Senate/Upgrades.js

```diff
@@ -0,0 +1,195 @@
+import React, { useCallback, useMemo, useState } from 'react';
+import { Grid, Text, P, LinkButton } from 'indigo-react';
+import * as azimuth from 'azimuth-js';
+
+import { useNetwork } from 'store/network';
+import { usePointCursor } from 'store/pointCursor';
+
+import * as need from 'lib/need';
+import useEthereumTransaction from 'lib/useEthereumTransaction';
+import { GAS_LIMITS } from 'lib/constants';
+import { useLocalRouter } from 'lib/LocalRouter';
+
+import ViewHeader from 'components/ViewHeader';
+import InlineEthereumTransaction from 'components/InlineEthereumTransaction';
+import View from 'components/View';
+import BridgeForm from 'form/BridgeForm';
+import convertToInt from 'lib/convertToInt';
+
+function useCastUpgradeVote() {
+  const { contracts } = useNetwork();
+  const { pointCursor } = usePointCursor();
+
+  const _contracts = need.contracts(contracts);
+  const _point = need.point(pointCursor);
+
+  return useEthereumTransaction(
+    useCallback(
+      (address, accept) => {
+        return azimuth.ecliptic.castUpgradeVote(
+          _contracts,
+          _point,
+          address,
+          accept
+        );
+      },
+      [_contracts, _point]
+    ),
+    useCallback(() => null, []), //TODO update polls[address].hasVoted ?
+    GAS_LIMITS.DEFAULT //TODO update once we have a decent sample size
+  );
+}
+
+export default function Upgrades() {
+  const { pop } = useLocalRouter();
+  const { contracts } = useNetwork();
+  const { pointCursor } = usePointCursor();
+
+  const _contracts = need.contracts(contracts);
+  const _point = convertToInt(need.point(pointCursor), 10);
+
+  const { construct, bind } = useCastUpgradeVote();
+
+  const [proposals, setProposals] = useState([]);
+  const [polls, setPolls] = useState({});
+  const [loading, setLoading] = useState(true);
+  const [votingOn, setVotingOn] = useState({});
+
+  //TODO update every second?
+  const now = Math.round(new Date().getTime() / 1000);
+
+  useMemo(async () => {
```

Comment:
What's with the use of useMemo for side effects instead of useEffect?

Reply:
Probably past me not knowing what he was doing and struggling with async useEffect? Fixed!